Implement support for various record types

mozzieongit opened the pull request:
TODO:
- ~~[ ] Figure out how to deal with `@` in HTTPS/SVCB target name ([cascade#190](https://github.com/nlnetlabs/cascade/issues/190))~~
- document (in code and where applicable) that parsing the @ symbol in SVCB/HTTPS RRs is not implemented (TODO moved to Issue #595)

Added types:
- [x] TLSA
- [x] OPENPGPKEY
- [x] SSHFP
- [x] IPSECKEY

Further planned:
- [x] SVCB (needed for nlnetlabs.nl)
- [x] HTTPS (because it's basically SVCB)
- CAA [RFC 8659](https://www.rfc-editor.org/rfc/rfc8659.html#name-syntax) (postponed)
- SMIMEA (because it's basically TLSA) [RFC 8162](https://datatracker.ietf.org/doc/html/rfc8162#page-4) (postponed)

There _was_ a FIXME in `src/base/iana/sshfp.rs`: "These types don't actually have a mnemonic, only a description."
- [x] How should we handle that? Use the mnemonics as they are in code right now, create a variant of inet_enum that doesn't have mnemonics, or remove the types all together?

I based the code on the `Zonemd` types, and don't know if there are functions that could be removed or should be added.

I tried adding a test in `test-data/zonefiles/` but get the error that de-serialization of bytes is not supported in YAML.
- [x] How should I add tests for the types, if at all?

Also, there are no (too short) length checks in the parser as I didn't find a direct answer in the RFCs, and didn't want to think too much about what a minimum would be given the possible parameters.

Fixes: #451
Fixes: #452
Fixes: #453
Fixes: #454

mozzieongit committed:
Implement support for SSHFP records

mozzieongit committed:
Implement support for OPENPGPKEY records

mozzieongit committed:
Implement support for TLSA records

mozzieongit committed:
Clippy

mozzieongit committed:
Resolve FIXME for SSHFP

mozzieongit committed:
Implement support for IPSECKEY records

mozzieongit committed:
Cargo clippy and fmt

mozzieongit committed:
Add tests for IPSECKEY

mozzieongit committed:
Add key length checks for IPSECKEY

mozzieongit committed:
Update comments and clippy

mozzieongit committed:
More clippy

mozzieongit committed:
Fix Debug of base64 key without std feature

mozzieongit committed:
Implement support for SVCB and HTTPS records in zonefiles

mozzieongit committed:
Cargo clippy

mozzieongit commented on src/rdata/svcb/value.rs:
```diff
@@ -608,6 +697,45 @@ impl<Octs: AsRef<[u8]>> Alpn<Octs> {
+                S::Error::custom("invalid svc param value for alpn")
+            })?))
+        } else {
+            // TODO: why is it ok none if the key is wrong? (stolen from parse_value)
+            Ok(None)
```
@partim I copied this behavior from parse_value. What was the reason to do this Result<Option<>> and check the key thing?

mozzieongit committed:
Fix reading SvcParamKey without value

Philip-NLnetLabs commented on src/rdata/svcb/params.rs:
```diff
@@ -186,6 +194,132 @@ impl<Octs: AsRef<[u8]>> SvcParams<Octs> {
+                .compose_value(&mut builder)
+                .map_err(|_| S::Error::short_buf())?;
+
+            // TODO: [...] Other automatically mandatory keys SHOULD NOT
```
In general the presentation format parser cannot know which keys are automatically mandatory. Except for the Mandatory key which is always automatically mandatory.

mozzieongit replied:
You are correct, but the TODO is in the context of parsing as SVCB and HTTPS, and when being parsed as an HTTPS RR additional mandatory keys are known.

mozzieongit commented on src/rdata/sshfp.rs:
```diff
@@ -0,0 +1,352 @@
+    fn fmt(&self, p: &mut impl Formatter) -> zonefile_fmt::Result {
+        p.block(|p| {
+            p.write_token(self.algorithm)?;
+            p.write_show(self.fingerprint_type)?;
```
Not sure, probably a left over from copying zonemd. I'll change the show to token.

mozzieongit committed:
Use write_token for SSHFP fptype

mozzieongit commented on src/rdata/ipseckey.rs:
```diff
@@ -0,0 +1,935 @@
+    /// Create a Ipseckey record data from provided parameters.
+    pub fn new(
+        precedence: u8,
+        gateway_type: IpseckeyGatewayType,
```
My initial thinking was to allow the user to do whatever they like, but now I think `domain` is designed to be easily correct for users, so yes, removing the type parameter is a good idea.

mozzieongit committed:
Derive ipseckey gateway type from gateway

mozzieongit committed:
Merge branch 'main' into add-rrtypes

mozzieongit committed:
Remove ipseckey gateway type parameter from tests

mozzieongit committed:
Update TODO to explicitly mention HTTPS RR

mozzieongit committed:
Cargo fmt

mozzieongit committed:
Merge branch 'main' into add-rrtypes

mozzieongit committed:
Add SVCB test vectors from RFC 9460

mozzieongit committed:
Fix error on missing SvcParam values

mozzieongit committed:
Make SVCB test rr easier to read

mozzieongit committed:
Disable SVCB alpn escaping

Implementing evaluating the escaped backslash and comma would require
access to a mutable octet sequence, but instead of implementing that
right now, we can just forbid escapes in alpn as permitted by the RFC.

mozzieongit committed:
Merge branch 'main' into add-rrtypes

mozzieongit committed:
Move SVCB zonefile tests into their own module with feature gate

mozzieongit committed:
Cargo fmt

mozzieongit committed:
Cargo clippy
